README.md
# 需要的module
web3，requests


# FollowMint

选一个狗王，跟着狗王，狗王mint啥软件自动帮你mint啥

已经自动排除非mint交易，非免费交易，已做模拟交易，模拟成功的才发送


编辑config.json里的参数：

**RPC**：ETH节点RPC

**privateKey**：私钥数组["私钥1", "私钥2"]，有几个号加几个

**blocknativeKey**：https://www.blocknative.com/  创建个账号，个人中心获取，不需要设置，直接复制key即可，和alchemyKey二选一，用哪个填哪个，不填置”“

**alchemyKey**：https://www.alchemy.com/ 创建ETH项目，获取，这个也是用于监控，和blocknative二选一，用哪个填哪个，不填置”“

**barkKey**：IOS的bark软件推送key，用于推送mint成功或者失败信息（没有可以空着）

**scanApikey**：https://etherscan.io/register 区块浏览器注册获取

**maxGasPrice**：最大的GAS，超过不会跟

**maxGasLimit**：最大的GAS消耗，超过不会跟

**maxValue**: 最大金额（设置0就只跟免费的，0.1就表示收费0.1以下的也跟）

**follow**：需要跟随的狗王地址，"follow": {"狗王地址":{"start": 开始跟单时间, "end": 结束跟单时间}}

**blacklist**：土狗名字黑名单，含有这些词的屏蔽

# 报错
1，async for' requires an object with __aiter__ method

  pip install websockets==10.3更新下websockets既可

2，handshake failed

  网络连接不上blocknative，尝试更换alchemy试试


# 已打包的exe
自行打包或者找朋友帮忙pyinstaller打包吧

实在要用exe的进discord下载吧

直接提示错误的，应该是配置文件错误

# 推特
https://twitter.com/fooyao158

# Discord

https://discord.gg/mgrcUfPEF9

# 打赏

fooyao.eth



changelog.md
2022年6月10日 8:37：

        1，更改数据判断，防止有人用合约批量撸的被识别为最大数量
        2，优化mint判断



2022年6月9日 17:17：

        1，修复wss断线重连的错误，
        2，30分钟无返回值，防止掉线，重新链接


2022年6月9日 11:21：

        1，增加maxMintNum参数，设置单包最大数量，超过数量的不打，可以过滤10个20个那种垃圾
        
        2，修复交易预估中一处逻辑错误
        
        3，已知wss一种证书问题，暂时不知道怎么解决


2022年6月8日 18:40：

        1，合并alpha版本
        
        2，监控功能新增alchemy，连接blocknative困难的可以尝试使用。配置文件alchemyKey和blocknativeKey俩个参数，哪个有值使用哪个监控方式，二选一
        
        3，maxFeePerGas改为2倍，防止在gas飙升的时候交易时间过长，但是可能会使gas上限超过设定，请择优选取狗王。
        
        4，监控wss自写，不再需要blocknative-sdk模块
        
        5，修复跟单开始结束时间判断的bug
        


FollowMint_win.py
import ctypes
import json
import sys
import threading
import time
import requests
from blocknative.stream import Stream
from web3 import Web3
from eth_abi import decode_abi
import os

configExample = {
    "RPC": "https://mainnet.infura.io/v3/9aa3d95b3bc440fa88ea12eaa4456161",
    "privateKey": ["私钥1", "私钥2"],
    "blocknativeKey": "监控平台key",
    "barkKey": "IOS推送软件key",
    "scanApikey": "https://etherscan.io/register注册获取",
    "maxGasPrice": 50,
    "maxGasLimit": 1000000,
    "maxValue": 0,
    "follow": {
        "0x8888887a5e2491fec904d90044e6cd6c69f1e71c": {"start": 0, "end": 24},
        "0x555555B63d1C3A8c09FB109d2c80464685Ee042B": {"start": 18, "end": 6},
        "0x99999983c70de9543cdc11dB5DE66A457d241e8B": {"start": 8, "end": 20}
    },
    "blacklist": ["Ape", "Bear", "Duck", "Pixel", "Not", "Okay", "Woman", "Baby", "Goblin", "Ai"]
}
std_out_handle = ctypes.windll.kernel32.GetStdHandle(-11)


def set_cmd_text_color(color, handle=std_out_handle):
    Bool = ctypes.windll.kernel32.SetConsoleTextAttribute(handle, color)
    return Bool


# reset white
def resetColor():
    set_cmd_text_color(0x0c | 0x0a | 0x09)


def print_green(message):
    stime = time.strftime("%m-%d %H:%M:%S", time.localtime())
    set_cmd_text_color(0x0a)
    sys.stdout.write(f'[{stime}] {message}\n')
    resetColor()


def print_red(message):
    stime = time.strftime("%m-%d %H:%M:%S", time.localtime())
    set_cmd_text_color(0x0c)
    sys.stdout.write(f'[{stime}] {message}\n')
    resetColor()


def print_blue(message):
    stime = time.strftime("%m-%d %H:%M:%S", time.localtime())
    set_cmd_text_color(0x0b)
    sys.stdout.write(f'[{stime}] {message}\n')
    resetColor()


def print_yellow(message):
    stime = time.strftime("%m-%d %H:%M:%S", time.localtime())
    set_cmd_text_color(0x0e)
    sys.stdout.write(f'[{stime}] {message}\n')
    resetColor()


def bark(info, data):
    if barkKey != '':
        requests.get('https://api.day.app/' + barkKey + '/' + info + '?url=' + data)


def getETHPrice():
    params = {
        'module': 'stats',
        'action': 'ethprice',
        'apikey': scanApikey
    }
    res = requests.get('https://api.etherscan.io/api', params=params)
    if res.status_code == 200 and res.json()['status'] == '1':
        return float(res.json()['result']['ethusd'])


def getMethodName(methodSignature, _conadd):
    try:
        if methodSignature in methodNameDict:
            print_blue('mint方法：' + methodNameDict[methodSignature]['method'])
            return methodNameDict[methodSignature]['isMint'], methodNameDict[methodSignature]['method']
        res = requests.get('https://www.4byte.directory/api/v1/signatures/?hex_signature=' + methodSignature)
        if res.status_code == 200 and res.json()['count'] > 0:
            method = res.json()['results'][0]['text_signature']
            methodName = method.split('(')[0].lower()
            print_blue('mint方法：' + method)
            if 'mint' in methodName:
                methodNameDict[methodSignature] = {'method': method, 'isMint': True}
                return True, method
            else:
                methodNameDict[methodSignature] = {'method': method, 'isMint': False}
        else:
            params = {
                'module': 'contract',
                'action': 'getabi',
                'address': _conadd,
                'apikey': scanApikey
            }
            res = requests.get('https://api.etherscan.io/api', params=params)
            if res.status_code == 200 and res.json()['status'] == '1':
                abi = res.json()['result']
                NFTcon = w3.eth.contract(address=w3.toChecksumAddress(_conadd), abi=abi)
                abiinfo = NFTcon.get_function_by_selector(methodSignature)
                method = str(abiinfo)[10:-1]
                print_blue('mint方法：' + method)
                if 'mint' in method.split('(')[0].lower():
                    methodNameDict[methodSignature] = {'method': method, 'isMint': True}
                    return True, method
                else:
                    methodNameDict[methodSignature] = {'method': method, 'isMint': False}
        return False, None
    except:
        return False, None


def isBlackList(_to):
    try:
        NFTcon = w3.eth.contract(address=_to, abi=[nameabi])
        name = NFTcon.functions.name().call()
        for black in blacklist:
            if black in name:
                print_red(name + "黑名单" + black + "，跳过")
                return False
    except:
        print_red('获取NFT名称失败，跳过')
        return False
    print_blue('NFT名称：' + name)
    return True


def isMintTime(_from):
    for _follow in follows:
        if _follow.lower() == _from.lower():
            starttime = int(follows[_follow]['start'])
            endtime = int(follows[_follow]['end'])
            tm_hour = time.localtime().tm_hour
            if tm_hour >= starttime or tm_hour < endtime:
                pass
            else:
                print_red("非Mint时间，跳过")
                return False
    return True


def getgas():
    headers = {"Authorization": blocknativeKey}
    url = 'https://api.blocknative.com/gasprices/blockprices?confidenceLevels=95'
    res = requests.get(url=url, headers=headers)
    if res.status_code == 200:
        estimatedPrices = res.json()['blockPrices'][0]['estimatedPrices'][0]
        maxPriorityFeePerGas = estimatedPrices['maxPriorityFeePerGas']
        maxFeePerGas = estimatedPrices['maxFeePerGas']
        baseFee = estimatedPrices['price']
        maxPriorityFeePerGas = w3.toWei(maxPriorityFeePerGas + 0.1, 'gwei')
        maxFeePerGas = w3.toWei(maxFeePerGas + 0.1, 'gwei')
        baseFee = w3.toWei(baseFee + 0.1, 'gwei')
        return baseFee, maxFeePerGas, maxPriorityFeePerGas


def minttx(_account, _privateKey, _inputData, _method, _from_address, _to_address, _maxFeePerGas, _maxPriorityFeePerGas, _value):
    try:
        abi = _method.split('(')[1][:-1].split(',')
        if len(abi) != 0 and 'address' in abi:
            params = decode_abi(abi, bytes.fromhex(_inputData[10:]))
            for index in range(len(abi)):
                if abi[index] == 'address':
                    _inputData = _inputData.replace(params[index][2:].lower(), _account.address[2:].lower())
        transaction = {
            'from': _account.address,
            'chainId': chainId,
            'to': _to_address,
            'gas': 2000000,
            'maxFeePerGas': _maxFeePerGas,
            'maxPriorityFeePerGas': _maxPriorityFeePerGas,
            'nonce': w3.eth.getTransactionCount(_account.address),
            'data': _inputData,
            'value': _value
        }
        try:
            estimateGas = w3.eth.estimateGas(transaction)
            if estimateGas > maxGasLimit:
                print_red('超过gasLimit上限，跳过')
                return
            transaction['gas'] = int(estimateGas * 1.2)
            signed = w3.eth.account.sign_transaction(transaction, _privateKey)
            new_raw = signed.rawTransaction.hex()
            tx_hash = w3.eth.sendRawTransaction(new_raw)
            print_green("mint交易发送成功" + w3.toHex(tx_hash))
            freceipt = w3.eth.waitForTransactionReceipt(tx_hash, 600)
            if freceipt.status == 1:
                try:
                    ETHused = freceipt.effectiveGasPrice * freceipt.gasUsed
                    ETHused = float(w3.fromWei(ETHused, 'ether'))
                    USDuse = ETHPrice * ETHused
                    ETHusedinfo = '本次mint：' + str(USDuse) + ' U'
                except Exception as e:
                    print(str(e))
                    ETHusedinfo = ''
                print_green("mint成功   " + ETHusedinfo)
                bark('mint成功', 'https://cn.etherscan.com/tx/' + w3.toHex(tx_hash))
            else:
                print_red("mint失败")
                if _to_address in mintadd:
                    mintadd.remove(_to_address)
                bark('mint失败', 'https://cn.etherscan.com/tx/' + w3.toHex(tx_hash))
        except Exception as e:
            print_red('预测失败，跳过:' + str(e))
            if _to_address in mintadd:
                mintadd.remove(_to_address)
            return
    except Exception as e:
        if _to_address in mintadd:
            mintadd.remove(_to_address)
        print_red('发送交易失败，跳过:' + str(e))
        return


async def txn_handler(txn, unsubscribe):
    to_address = txn['to']
    from_address = txn['from']
    to_address = w3.toChecksumAddress(to_address)
    inputData = txn['input']
    value = int(txn['value'])
    print_yellow(from_address + "监控到新交易")
    if not isMintTime(from_address):
        return
    if value > maxValue:
        print_red("超过最大金额，跳过")
        return
    if to_address in mintadd:
        print_red("mint过，跳过")
        return
    isMint, method = getMethodName(inputData[:10], to_address)
    if not isMint:
        print_red('可能不是mint交易,跳过')
        return
    if not isBlackList(to_address):
        return
    gasPrice, maxFeePerGas, maxPriorityFeePerGas = getgas()
    if gasPrice > maxGasPrice:
        print_red('gasPrice过高,跳过')
        return
    mintadd.append(to_address)
    for index in range(len(accounts)):
        threading.Thread(target=minttx, args=(accounts[index], privateKeys[index], inputData, method, from_address, to_address, maxFeePerGas, maxPriorityFeePerGas, value)).start()


def main():
    while True:
        try:
            stream = Stream(blocknativeKey)
            print_blue(str(len(accounts)) + '个地址开始监控')
            print_blue('开始监控')
            for _follow in follows:
                filters = [{
                    "status": "pending",
                    "from": _follow
                }]
                stream.subscribe_address(_follow, txn_handler, filters)
            stream.connect()
        except Exception as e:
            print_red(str(e))
            time.sleep(10)


if __name__ == '__main__':
    print_red("有能力的请使用源码，不对使用者安全负责")
    print_red("打狗请用小号，无法保证无bug")
    print_red("开源地址：https://github.com/Fooyao/FollowMint")
    print_red("代码水平较差，有任何优化建议请反馈")
    if not os.path.exists('config.json'):
        print_blue('请先配置config.json')
        file = open('config.json', 'w')
        file.write(json.dumps(configExample))
        file.close()
        time.sleep(10)
    try:
        file = open('config.json', 'r')
        config = json.loads(file.read())
        RPC = config['RPC']
        privateKeys = config['privateKey']
        scanApikey = config['scanApikey']
        blacklist = config['blacklist']
        maxValue = config['maxValue']
        blocknativeKey = config['blocknativeKey']
        barkKey = config['barkKey']
        follows = config['follow']
        nameabi = {
            'inputs': [],
            'name': 'name',
            'outputs': [{'internalType': 'string', 'name': '', 'type': 'string'}],
            'stateMutability': 'view',
            'type': 'function'
        }
        w3 = Web3(Web3.HTTPProvider(RPC))
        maxGasPrice = config['maxGasPrice']
        maxGasPrice = w3.toWei(maxGasPrice, 'gwei')
        maxGasLimit = int(config['maxGasLimit'])
        maxValue = w3.toWei(maxValue, 'ether')
        chainId = w3.eth.chainId
        accounts = []
        for privateKey in privateKeys:
            accounts.append(w3.eth.account.privateKeyToAccount(privateKey))
        mintadd = []
        ETHPrice = getETHPrice()
        print('ETH单价: ' + str(ETHPrice))
        methodNameDict = {}
        main()
    except Exception as e:
        print_red(str(e))
        time.sleep(10)


FollowMint.py
import os
import json
import time
import asyncio
import platform
import requests
import threading
import websockets
from web3 import Web3
from datetime import datetime
from eth_abi import decode_abi

# 这里是作为模板输出到config用，修改无用
configExample = {
    "RPC": "https://mainnet.infura.io/v3/9aa3d95b3bc440fa88ea12eaa4456161",  # RPC节点，得支持w3.eth.filter才行
    "privateKey": ["privateKey1", "privateKey2"],  # 私钥，支持多个
    "blocknativeKey": "",  # blocknative的key，和alchemyKey二选一即可
    "alchemyKey": "",  # alchemy的key，和blocknativeKey二选一即可
    "barkKey": "",
    "scanApikey": "",
    "maxGasPrice": 50,  # baseGas上限，超过不打
    "maxGasLimit": 1000000,
    "maxValue": 0,  # 0表示只打免费，0.1表示付费金额小于0.1ETH的都打
    "maxMintNum": 5,  # 只打一个包5个NFT以内的
    "follow": {
        "0x8888887a5e2491fec904d90044e6cd6c69f1e71c": {"start": 0, "end": 24},   # 0点到24点全天跟单
        "0x555555B63d1C3A8c09FB109d2c80464685Ee042B": {"start": 18, "end": 6},
        "0x99999983c70de9543cdc11dB5DE66A457d241e8B": {"start": 8, "end": 20}
    },
    "blacklist": ["Ape", "Pixel", "Not", "Okay", "Woman", "Baby", "Goblin", "Ai"]  # 黑名单区分大小写
}

# 输出颜色优化
if platform.system().lower() == 'windows':
    import ctypes
    import sys

    std_out_handle = ctypes.windll.kernel32.GetStdHandle(-11)


    def set_cmd_text_color(color, handle=std_out_handle):
        Bool = ctypes.windll.kernel32.SetConsoleTextAttribute(handle, color)
        return Bool


    def print_color(message, color):
        colorDict = {'green': 0x0a, 'red': 0x0c, 'blue': 0x0b, 'yellow': 0x0e}
        stime = time.strftime("%m-%d %H:%M:%S", time.localtime())
        set_cmd_text_color(colorDict[color])
        sys.stdout.write(f'[{stime}] {message}\n')
        set_cmd_text_color(0x0c | 0x0a | 0x09)
else:
    def print_color(message, color):
        colorDict = {'green': '32m', 'red': '31m', 'blue': '34m', 'yellow': '33m'}
        stime = time.strftime("%m-%d %H:%M:%S", time.localtime())
        print(f'[{stime}] \033[1;{colorDict[color]}{message}\033[0m')


# bark推送
def bark(info, data):
    if barkKey != '':
        requests.get('https://api.day.app/' + barkKey + '/' + info + '?url=' + data)


def getETHPrice():
    params = {
        'module': 'stats',
        'action': 'ethprice',
        'apikey': scanApikey
    }
    res = requests.get('https://api.etherscan.io/api', params=params)
    if res.status_code == 200 and res.json()['status'] == '1':
        return float(res.json()['result']['ethusd'])


def getMethodName(_inputData, _conadd, _pendingBlockNumber):
    try:
        isMint, mintNum, safeMint = isMintFromBlock(_pendingBlockNumber, _conadd, _inputData)
        if isMint:
            if safeMint:
                return True, None, mintNum, True
            methodSignature = _inputData[:10]
            if methodSignature in methodNameDict:
                print_color('mint方法：' + methodNameDict[methodSignature]['method'], 'blue')
                return methodNameDict[methodSignature]['isMint'], methodNameDict[methodSignature]['method'], mintNum, False
            res = requests.get('https://www.4byte.directory/api/v1/signatures/?hex_signature=' + methodSignature)
            if res.status_code == 200 and res.json()['count'] > 0:
                method = res.json()['results'][0]['text_signature']
                print_color('mint方法：' + method, 'blue')
                methodNameDict[methodSignature] = {'method': method, 'isMint': True}
                return True, method, mintNum, False
            else:
                params = {
                    'module': 'contract',
                    'action': 'getabi',
                    'address': _conadd,
                    'apikey': scanApikey
                }
                res = requests.get('https://api.etherscan.io/api', params=params)
                if res.status_code == 200 and res.json()['status'] == '1':
                    abi = res.json()['result']
                    NFTcon = w3.eth.contract(address=w3.toChecksumAddress(_conadd), abi=abi)
                    abiinfo = NFTcon.get_function_by_selector(methodSignature)
                    method = str(abiinfo)[10:-1]
                    print_color('mint方法：' + method, 'blue')
                    methodNameDict[methodSignature] = {'method': method, 'isMint': True}
                    return True, method, mintNum, False
        else:
            return False, None, 0, False
    except:
        return False, None, 0, False


def isMintFromBlock(_block, _conadd, _inputData):
    try:
        log = w3.eth.filter({
            'fromBlock': _block - 11,
            'toBlock': _block - 1,
            'address': w3.toChecksumAddress(_conadd),
            'topics': ['0xddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3ef',
                       '0x0000000000000000000000000000000000000000000000000000000000000000']
        })  # 拉取所有0地址转移出来NFT的logs
        txs = log.get_all_entries()
        if len(txs) > 1:
            transactionHashDict = {}
            for tx in txs:
                if tx.transactionHash in transactionHashDict:
                    transactionHashDict[tx.transactionHash] += 1
                else:
                    transactionHashDict[tx.transactionHash] = 1
            transactionHashList = list(transactionHashDict.values())
            allMaxMint = max(transactionHashList, key=transactionHashList.count)  # 获取重复交易哈希出现最多次数，也就是最大mint次数
            transaction = w3.eth.get_transaction(txs[0]['transactionHash'])
            _input = transaction['input'][:10]
            if _inputData == _input:
                return True, allMaxMint, True
            elif _inputData[:10] == _input[:10]:
                return True, allMaxMint, False
    except Exception as e:
        print(str(e))
    return False, 0, False


def isBlackList(_to):
    try:
        NFTcon = w3.eth.contract(address=_to, abi=[nameabi])
        name = NFTcon.functions.name().call()
        for black in blacklist:
            if black in name:
                print_color(name + "黑名单" + black + "，跳过", 'red')
                return False
    except:
        print_color('获取NFT名称失败，跳过', 'red')
        return False
    print_color('NFT名称：' + name, 'blue')
    return True


def isMintTime(_from):
    if _from.lower() in follows:
        starttime = int(follows[_from.lower()]['start'])
        endtime = int(follows[_from.lower()]['end'])
        tm_hour = time.localtime().tm_hour
        if starttime < endtime:
            if not starttime <= tm_hour < endtime:
                print_color("非Mint时间，跳过", 'red')
                return False
        else:
            if endtime <= tm_hour < starttime:
                print_color("非Mint时间，跳过", 'red')
                return False
    return True


def getgas():
    url = 'https://blocknative-api.herokuapp.com/data'
    res = requests.get(url=url)
    if res.status_code == 200:
        estimatedPrices = res.json()['estimatedPrices'][0]
        maxPriorityFeePerGas, maxFeePerGas, baseFee = estimatedPrices['maxPriorityFeePerGas'], estimatedPrices['maxFeePerGas'], estimatedPrices['price']
        maxPriorityFeePerGas, maxFeePerGas, baseFee = w3.toWei(maxPriorityFeePerGas + 0.1, 'gwei'), w3.toWei(maxFeePerGas + 0.1, 'gwei'), w3.toWei(baseFee + 0.1, 'gwei')
        return baseFee, maxFeePerGas, maxPriorityFeePerGas


def minttx(_account, _privateKey, _inputData, _method, _safeMint, _from_address, _to_address, _maxFeePerGas, _maxPriorityFeePerGas, _value, _gasLimit):
    try:
        if not _safeMint:
            abi = _method.split('(')[1][:-1].split(',')
            if len(abi) != 0 and 'address' in abi:
                params = decode_abi(abi, bytes.fromhex(_inputData[10:]))
                for index in range(len(abi)):
                    if abi[index] == 'address':
                        _inputData = _inputData.replace(params[index][2:].lower(), _account.address[2:].lower())
        transaction = {
            'from': _account.address,
            'chainId': chainId,
            'to': _to_address,
            'gas': _gasLimit,
            'maxFeePerGas': int(_maxFeePerGas * 2),
            'maxPriorityFeePerGas': _maxPriorityFeePerGas,
            'nonce': w3.eth.getTransactionCount(_account.address),
            'data': _inputData,
            'value': _value
        }
        try:
            estimateGas = w3.eth.estimateGas(transaction)
            if estimateGas > _gasLimit:
                transaction['gas'] = int(estimateGas * 1.2)
        except Exception as e:
            if 'max fee per gas less than block base fee' in str(e):
                _, maxFeePerGas, maxPriorityFeePerGas = getgas()
                transaction['maxFeePerGas'] = int(maxFeePerGas * 2)
                transaction['maxPriorityFeePerGas'] = maxPriorityFeePerGas
            elif 'allowance' not in str(e):
                print_color('发送交易失败:' + str(e), 'red')
                if _to_address in mintadd:
                    mintadd.remove(_to_address)
                return
        signed = w3.eth.account.sign_transaction(transaction, _privateKey)
        new_raw = signed.rawTransaction.hex()
        tx_hash = w3.eth.sendRawTransaction(new_raw)
        print_color("mint交易发送成功" + w3.toHex(tx_hash), 'green')
        freceipt = w3.eth.waitForTransactionReceipt(tx_hash, 600)
        if freceipt.status == 1:
            try:
                ETHused = freceipt.effectiveGasPrice * freceipt.gasUsed
                ETHused = float(w3.fromWei(ETHused, 'ether'))
                USDuse = ETHPrice * ETHused
                ETHusedinfo = '本次mint：' + str(USDuse) + ' U'
            except:
                ETHusedinfo = ''
            print_color("mint成功   " + ETHusedinfo, 'green')
            bark('mint成功', 'https://cn.etherscan.com/tx/' + w3.toHex(tx_hash))
        else:
            print_color("mint失败", 'red')
            if _to_address in mintadd:
                mintadd.remove(_to_address)
            bark('mint失败', 'https://cn.etherscan.com/tx/' + w3.toHex(tx_hash))
    except Exception as e:
        if _to_address in mintadd:
            mintadd.remove(_to_address)
        print_color('交易失败:' + str(e), 'red')
        return


def txn_handler(to_address, from_address, inputData, value, gasLimit, tx_maxFeePerGas, tx_maxPriorityFeePerGas, pendingBlockNumber):
    to_address = w3.toChecksumAddress(to_address)
    print_color(from_address + "监控到新交易", 'yellow')
    if not isMintTime(from_address):
        return
    if value > maxValue:
        print_color("超过最大金额，跳过", 'red')
        return
    if to_address in mintadd:
        print_color("mint过，跳过", 'red')
        return
    isMint, method, mintNum, safeMint = getMethodName(inputData, to_address, pendingBlockNumber)
    if not isMint:
        print_color('可能不是mint交易,跳过', 'red')
        return
    if mintNum > maxMintNum:
        print_color(f'mint数量{mintNum}超过设定上限{maxMintNum},跳过', 'red')
        return
    if not isBlackList(to_address):
        return
    gasPrice, maxFeePerGas, maxPriorityFeePerGas = getgas()
    if gasPrice > maxGasPrice:
        print_color('gasPrice过高,跳过', 'red')
        return
    maxFeePerGas = maxFeePerGas if maxFeePerGas > tx_maxFeePerGas else tx_maxFeePerGas
    maxPriorityFeePerGas = maxPriorityFeePerGas if maxPriorityFeePerGas > tx_maxPriorityFeePerGas else tx_maxPriorityFeePerGas
    if gasPrice + maxPriorityFeePerGas < maxFeePerGas:
        maxFeePerGas = gasPrice + maxPriorityFeePerGas
    mintadd.append(to_address)
    for index in range(len(accounts)):
        threading.Thread(target=minttx, args=(accounts[index], privateKeys[index], inputData, method, safeMint, from_address, to_address, maxFeePerGas, maxPriorityFeePerGas, value, gasLimit)).start()


async def blocknative():
    async for websocket in websockets.connect('wss://api.blocknative.com/v0'):
        try:
            initialize = {
                "categoryCode": "initialize", "eventCode": "checkDappId",
                "dappId": blocknativeKey, "timeStamp": datetime.utcnow().strftime('%Y-%m-%dT%H:%M:%S.%f')[:-3] + 'Z',
                "version": "1",
                "blockchain": {"system": "ethereum", "network": "main"}
            }
            await websocket.send(json.dumps(initialize))
            for _follow in follows:
                configs = {
                    "categoryCode": "configs", "eventCode": "put",
                    "config": {"scope": _follow, "filters": [{"from": _follow, "status": "pending"}], "watchAddress": True},
                    "dappId": blocknativeKey,
                    "timeStamp": datetime.utcnow().strftime('%Y-%m-%dT%H:%M:%S.%f')[:-3] + 'Z',
                    "version": "1",
                    "blockchain": {"system": "ethereum", "network": "main"}
                }
                await websocket.send(json.dumps(configs))
            while True:
                try:
                    message = await asyncio.wait_for(websocket.recv(), timeout=30*60)
                except asyncio.TimeoutError:
                    print_color('30分钟无消息，可能断开，尝试重连', 'red')
                    await websocket.close()
                    break
                json_data = json.loads(message)
                if json_data['status'] == 'ok' and 'event' in json_data:
                    if json_data['event']['categoryCode'] == 'initialize':
                        print_color('初始化成功', 'blue')
                    elif json_data['event']['categoryCode'] == 'configs':
                        print_color(f"监控{json_data['event']['config']['scope']}地址成功", 'blue')
                    elif json_data['event']['categoryCode'] == 'activeAddress':
                        txn = json_data['event']['transaction']
                        to_address, from_address, inputData, gasLimit, value = txn['to'], txn['from'], txn['input'], int(txn['gas']), int(txn['value'])
                        if 'maxFeePerGas' in txn:
                            tx_maxFeePerGas, tx_maxPriorityFeePerGas = int(txn['maxFeePerGas']), int(txn['maxPriorityFeePerGas'])
                        else:
                            tx_maxFeePerGas, tx_maxPriorityFeePerGas = 0, 0
                        pendingBlockNumber = int(txn['pendingBlockNumber'])
                        threading.Thread(target=txn_handler, args=(to_address, from_address, inputData, value, gasLimit, tx_maxFeePerGas, tx_maxPriorityFeePerGas, pendingBlockNumber)).start()
                    else:
                        print_color(message, 'blue')
        except Exception as e:
            print_color(str(e), 'red')
            await websocket.close()


async def alchemy():
    async for websocket in websockets.connect(f'wss://eth-mainnet.alchemyapi.io/v2/{alchemyKey}'):
        try:
            json_data = {
                "jsonrpc": "2.0",
                "id": 1, "method": "eth_subscribe", "params": []
            }
            for _follow in follows:
                json_data['params'] = ["alchemy_filteredNewFullPendingTransactions", {"address": _follow}]
                await websocket.send(json.dumps(json_data))
                result = await websocket.recv()
                if "result" in result:
                    print_color(f"监控{_follow}地址成功", 'blue')
            while True:
                try:
                    message = await asyncio.wait_for(websocket.recv(), timeout=30*60)
                except asyncio.TimeoutError:
                    print_color('30分钟无消息，可能断开，尝试重连', 'red')
                    await websocket.close()
                    break
                json_data = json.loads(message)
                if 'params' in json_data:
                    txn = json_data['params']['result']
                    to_address, from_address, inputData, gasLimit, value = txn['to'], txn['from'], txn['input'], int(txn['gas'], 16), int(txn['value'], 16)
                    if from_address.lower() not in follows:
                        return
                    if 'maxFeePerGas' in txn:
                        tx_maxFeePerGas, tx_maxPriorityFeePerGas = int(txn['maxFeePerGas'], 16), int(txn['maxPriorityFeePerGas'], 16)
                    else:
                        tx_maxFeePerGas, tx_maxPriorityFeePerGas = 0, 0
                    pendingBlockNumber = w3.eth.get_block_number()
                    threading.Thread(target=txn_handler, args=(to_address, from_address, inputData, value, gasLimit, tx_maxFeePerGas, tx_maxPriorityFeePerGas, pendingBlockNumber)).start()
        except Exception as e:
            print_color(str(e), 'red')
            await websocket.close()


if __name__ == '__main__':
    print_color("有能力的请使用源码，不对使用者安全负责", 'red')
    print_color("打狗请用小号，无法保证无bug", 'red')
    print_color("开源地址：https://github.com/Fooyao/FollowMint", 'red')
    print_color("代码水平较差，有任何优化建议请反馈", 'red')
    if not os.path.exists('config.json'):
        print_color('请先配置config.json', 'blue')
        file = open('config.json', 'w')
        file.write(json.dumps(configExample))
        file.close()
        time.sleep(10)
    try:
        file = open('config.json', 'r')
        config = json.loads(file.read())
        RPC, privateKeys, scanApikey, blacklist = config['RPC'], config['privateKey'], config['scanApikey'], config['blacklist']
        blocknativeKey, alchemyKey, barkKey, follows = config['blocknativeKey'], config['alchemyKey'], config['barkKey'], config['follow']
        follows = dict((k.lower(), v) for k, v in follows.items())
        nameabi = {
            'inputs': [], 'name': 'name',
            'outputs': [{'internalType': 'string', 'name': '', 'type': 'string'}],
            'stateMutability': 'view', 'type': 'function'
        }
        w3 = Web3(Web3.HTTPProvider(RPC))
        maxGasPrice, maxGasLimit, maxValue, maxMintNum = config['maxGasPrice'], int(config['maxGasLimit']), config['maxValue'], int(config['maxMintNum'])
        maxGasPrice = w3.toWei(maxGasPrice, 'gwei')
        maxValue = w3.toWei(maxValue, 'ether')
        chainId = w3.eth.chainId
        accounts = [w3.eth.account.privateKeyToAccount(privateKey) for privateKey in privateKeys]
        mintadd = []
        ETHPrice = getETHPrice()
        print_color('ETH单价: ' + str(ETHPrice), 'green')
        methodNameDict = {}
        if len(blocknativeKey) >= 20:
            asyncio.run(blocknative())
        elif len(alchemyKey) >= 20:
            asyncio.run(alchemy())
        else:
            print_color('blocknativeKey和alchemyKey必须提供一个', 'red')
    except Exception as e:
        print_color(str(e), 'red')
        time.sleep(10)
